drivers/regulator/regulator_fixed.c

```diff
@@ -0,0 +1,295 @@
+	}
+
+	notify(mgr, rc);
+}
+
+static void stop(struct onoff_manager *mgr,
+		 onoff_notify_fn notify)
+{
+	const struct driver_data_onoff *data =
+		CONTAINER_OF(mgr, struct driver_data_onoff, mgr);
+	const struct driver_config *cfg = data->dev->config_info;
+	uint32_t delay_us = cfg->off_on_delay_us;
+	int rc;
+
+	if ((cfg->options & OPTION_ALWAYS_ON) != 0) {
+		rc = -EPERM;
+		goto out;
+	}
+
+	rc = gpio_pin_set(data->gpio, cfg->gpio_pin, false);
+
+	if (rc == -EWOULDBLOCK) {
+		/* TODO: delegate completion to a work queue. */
+	} else if ((rc >= 0) && (delay_us > 0)) {
+		/* Technically we could return immediately and just increase the
+		 * delay on the next start if it's requested before the required
+		 * delay completes.  That would require caching the time we
+		 * turned the regulator off, which is too costly.
+		 */
+		common_delay(delay_us);
+	}
+out:
+	notify(mgr, rc);
+}
+
+static int enable_onoff(struct device *dev, struct onoff_client *cli)
+{
+	struct driver_data_onoff *data = dev->driver_data;
+
+	return onoff_request(&data->mgr, cli);
+}
+
+static int disable_onoff(struct device *dev)
+{
+	struct driver_data_onoff *data = dev->driver_data;
+
+	return onoff_release(&data->mgr);
+}
+
+static const struct onoff_transitions transitions =
+	ONOFF_TRANSITIONS_INITIALIZER(start, stop, NULL);
+
+static const struct regulator_driver_api api_onoff = {
+	.enable = enable_onoff,
+	.disable = disable_onoff,
+};
+
+static int regulator_fixed_init_onoff(struct device *dev)
+{
+	struct driver_data_onoff *data = dev->driver_data;
+	int rc = common_init(dev, &data->gpio);
+
+	data->dev = dev;
+	onoff_manager_init(&data->mgr, &transitions);
+
+	/* If this was turned on at boot then we need to go through the onoff
+	 * request to set the state, even though the signal is already
+	 * active.
+	 */
+	if (rc > 0) {
+		struct onoff_client cli;
+
+		sys_notify_init_spinwait(&cli.notify);
+
+		rc = enable_onoff(dev, &cli);
+
+		LOG_DBG("%s boot-on: %d", dev->name, rc);
+		if (rc >= 0) {
+			while (sys_notify_fetch_result(&cli.notify, &rc)
+			       == -EAGAIN) {
+				/* spin */
+			}
+		}
+	}
+	LOG_INF("%s onoff: %d", dev->name, rc);
+
+	return rc;
+}
+
+struct driver_data_sync {
+	struct device *gpio;
+	struct onoff_sync_service srv;
+};
+
+static int enable_sync(struct device *dev, struct onoff_client *cli)
+{
+	struct driver_data_sync *data = dev->driver_data;
+	const struct driver_config *cfg = dev->config_info;
+	k_spinlock_key_t key;
+	int rc = onoff_sync_lock(&data->srv, &key);
+
+	if (rc == 0) {
+		rc = gpio_pin_set(data->gpio, cfg->gpio_pin, true);
+	}
+
+	return onoff_sync_finalize(&data->srv, key, cli, rc, true);
+}
+
+static int disable_sync(struct device *dev)
+{
+	struct driver_data_sync *data = dev->driver_data;
+	const struct driver_config *cfg = dev->config_info;
+	k_spinlock_key_t key;
+	int rc = onoff_sync_lock(&data->srv, &key);
+
+	if ((rc == 1) && ((cfg->options & OPTION_ALWAYS_ON) != 0)) {
+		rc = -EPERM;
+	} else if (rc > 0) {
```

Comment:
shouldn't it be `rc == 1` so that only the last disabling user is cleaning the gpio?

Reply:
Reworked, thanks.